File: C_version/Ant.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <random>
#include <set>
#include <map>
#include "Ant.h"
#include "Environment.h"

using namespace std;

Ant::Ant(double _alpha, double _beta, int initial_location) 
    : alpha(_alpha), beta(_beta), current_location(initial_location), travelled_distance(0.0) {}

vector<vector<double>> Ant::precompute_probability_matrix() {
    const auto& edges = environment->edges;

    vector<vector<double>> probabilities(environment->n, vector<double>(environment->n, 0.0));

    // the probability of going from i to j is given by the formula
    // p_ij = (pheromone_jj ^alpa) * (1/d_ij ^ beta) / sum(pheromone_jj ^ alpha * 1/d_ij ^ beta)
    double sum = 0.0;
    for (auto&edge :edges) {
        double pheromone = edge.pheromone;
        double distance = 1.0 / get_distance(edge.from - 1, edge.to - 1);
        sum += pow(pheromone, alpha) * pow(distance, beta);
    }

    for (int i = 0; i < environment->n; ++i) {
        for (int j = 0; j < environment->n; ++j) {
            double pheromone = edges[i].pheromone;
            double distance = 1.0 / get_distance(i, j);
            probabilities[i][j] = (pow(pheromone, alpha) * pow(distance, beta)) / sum;
        }
    }

    return probabilities;
}

void Ant::run() {
    // clear visited locations and edges
    this->visited_locations.clear();
    this->visited_edges.clear();
    this->travelled_distance = 0.0;
    // add initial location to visited locations
    visited_locations.push_back(current_location);

    vector<int> not_yet_visited;

    for (int i = 1; i < environment->n + 1; ++i) {
        // if i is not in visited_locations, add it to not_yet_visited
        if (find(visited_locations.begin(), visited_locations.end(), i) != visited_locations.end()) continue;
        not_yet_visited.push_back(i);
    }

    auto probabilities = precompute_probability_matrix();

    while (visited_locations.size() < 48) {
        int next_location = select_path(not_yet_visited, probabilities);
        travelled_distance += get_distance(current_location - 1, next_location - 1);
        visited_locations.push_back(next_location);
        visited_edges.emplace_back(current_location, next_location);

        current_location = next_location;
        not_yet_visited.erase(remove(not_yet_visited.begin(), not_yet_visited.end(), next_location), not_yet_visited.end());

    }
    // Return to the initial location
    travelled_distance += get_distance(visited_locations.back() - 1, visited_locations.front() - 1);
    visited_edges.emplace_back(visited_locations.back(), visited_locations.front());
    visited_locations.push_back(visited_locations.front());
}

double Ant::get_distance(int i, int j) {
    return environment->get_distance(i, j);
}



int Ant::select_path(const vector<int>& not_yet_visited, const vector<vector<double>>& probabilities) {
    // from the probabilites matrix choose the vector corresponding to the current location
    // and select the next location based on the probabilities
    vector<int> choices = not_yet_visited;
    vector<double> probabilities_for_current_location = probabilities[current_location - 1];
    vector<double> probabilities_for_choices;

    for (int i = 0; i < choices.size(); ++i) {
        probabilities_for_choices.push_back(probabilities_for_current_location[choices[i] - 1]);
    }
    
    // normalize the probabilities
    double sum = 0.0;
    for (auto& prob : probabilities_for_choices) {
        sum += prob;
    }

    vector<double> normalized_probabilities;
    for (auto& prob : probabilities_for_choices) {
        normalized_probabilities.push_back(prob / sum);
    }

    // select the next location based on the probabilities
    random_device rd;
    mt19937 gen(rd());
    uniform_real_distribution<double> dis(0.0, 1.0);
    double random_number = dis(gen);

    double cumulative_probability = 0.0;
    for (int i = 0; i < normalized_probabilities.size(); ++i) {
        cumulative_probability += normalized_probabilities[i];
        if (random_number <= cumulative_probability) {
            return choices[i];
        }
    }
    return choices.back();
}

void Ant::join(Environment* env) {
    environment = env;
}

File: Py_version/ant.py
import tsplib95
import numpy as np

# Class representing an artificial ant of the ant colony
"""
    alpha: a parameter controlling the influence of the amount of pheromone during ants' path selection process
    beta: a parameter controlling the influence of the distance to the next node during ants' path selection process
"""
class Ant():
    def __init__(self, alpha: float, beta: float, initial_location: int):
        self.alpha = alpha
        self.beta = beta
        self.current_location = initial_location + 1
        self.travelled_distance = 0
        self.visited_locations = []
        self.visited_edges = []
        self.environment = None


    def precompute_probability_matrix(self):
        initial_pheromones = self.environment.pheromone_map
        intial_distances = self.environment.distances
    
        # Avoid division by zero: replace zeros in distances with a small value
        safe_distances = np.where(intial_distances == 0, np.inf, intial_distances)

        # this calculates the probabilities based on the ACO algorithm
        # the probability of moving from node i to node j is given by:
        # (pheromone[i, j] ** alpha) * ((1 / distance[i, j]) ** beta)
        probabilities = (initial_pheromones ** self.alpha) * ((1 / safe_distances) ** self.beta)


        return probabilities
        
   
    # The ant runs to visit all the possible locations of the environment 
    def run(self):
        # list of all nodes
        all_cities = [i+1 for i in range(self.environment.n)]
        # precalculate the pheromone matrix
        self.probabilities = self.precompute_probability_matrix()

        # Initialize the ant by visiting the initial location
        self.visited_locations = [self.current_location]
        self.travelled_distance = 0

        # not yet visited nodes
        not_yet_visited = set(all_cities) - set(self.visited_locations)

        while not_yet_visited:
            next_location = self.select_path(list(not_yet_visited))
            self.travelled_distance += self.get_distance(self.current_location, next_location)
            self.visited_locations.append(next_location)
            self.visited_edges.append((self.current_location, next_location))

            # Move to the next location
            self.current_location = next_location
            not_yet_visited.remove(next_location)

        # Return to the initial location
        self.travelled_distance += self.get_distance(self.visited_locations[-1], self.visited_locations[0])
        self.visited_edges.append((self.visited_locations[-1], self.visited_locations[0]))
        self.visited_locations.append(self.visited_locations[0])
        
                
    # Select the next path based on the random proportional rule of the ACO algorithm
    def select_path(self, not_yet_visited):
        # gives us the probability of moving to each node as a vector
        prob_vector = self.probabilities[self.current_location, not_yet_visited]
        total = np.sum(prob_vector)
        normalized_probabilities = prob_vector / total if total > 0 else prob_vector
        return np.random.choice(not_yet_visited, p=normalized_probabilities)
    

    # Position an ant in an environment
    def join(self, environment):
        self.environment = environment


    def get_distance(self, i, j):
        return self.environment.distances[i, j]

